src/iframe/storage/unpartitioned-storage/unpartitioned-storage.ts

```diff
@@ -182,31 +162,66 @@ export class EnhancedStorage implements Storage {
   }
 
   /**
-   * Set up a handler to request storage access on user interaction
-   * This is required by the Storage Access API for security
+   * Returns the current PermissionState and starts listening for permission changes, making sure only one listener is added.
    */
-  protected setupUserInteractionHandler(): void {
-    log(LOG_GROUP.STORAGE, "Waiting for user interaction to request storage access");
+  private async setupPermissionChangeHandlerAndGetCurrentValue(): Promise<PermissionState> {
+    let permissionStatus = this.permissionStatusPromise ? await this.permissionStatusPromise : null;
+    let permissionState: PermissionState = permissionStatus?.state;
 
-    // Create a reusable handler function
-    const handleUserInteraction = async () => {
+    if (navigator.permissions && !permissionStatus) {
       try {
-        await this.requestStorageAccess();
+        this.permissionStatusPromise = navigator.permissions.query({
+          name: "storage-access" as PermissionName,
+        });
+
+        permissionStatus = await this.permissionStatusPromise;
+        permissionState = permissionStatus.state;
 
-        log(LOG_GROUP.STORAGE, "Storage access granted after user interaction");
+        permissionStatus.addEventListener("change", () => {
```

Comment:
Before, we would only listen for permission changes if `document.hasStorageAccess()` returned `false`. Now we always do, because the permissions can change in any direction.